File: dsa-learning/dsa-python-code/com/dsa/leetcode/labuladong2/list/P707_DesignLinkedList.py
class Node:
    def __init__(self, val: int):
        self.val = val
        self.prev = None
        self.next = None


class MyLinkedList:

    def __init__(self):
        self.head = Node(None)
        self.tail = Node(None)
        self.size = 0
        self.head.next = self.tail
        self.tail.prev = self.head

    def get(self, index: int) -> int:
        if self.isNotPositionIndex(index):
            return -1
        return self.getNode(index).val

    def addAtHead(self, val: int) -> None:
        self.addAtIndex(0, val)

    def addAtTail(self, val: int) -> None:
        self.addAtIndex(self.size, val)

    def addAtIndex(self, index: int, val: int) -> None:
        if self.isPositionIndex(index):
            cur = self.getNode(index)
            prev = cur.prev
            newNode = Node(val)
            prev.next = newNode
            cur.prev = newNode
            newNode.prev = prev
            newNode.next = cur
            self.size = self.size + 1
        if index == self.size:
            cur = self.tail
            prev = cur.prev
            newNode = Node(val)
            prev.next = newNode
            cur.prev = newNode
            newNode.prev = prev
            newNode.next = cur
            self.size = self.size + 1

    def deleteAtIndex(self, index: int) -> None:
        if self.isPositionIndex(index):
            cur = self.getNode(index)
            prev = cur.prev
            next = cur.next
            prev.next = next
            next.prev = prev
            self.size = self.size - 1

    def getNode(self, index: int) -> Node:
        cur = self.head
        for i in range(0, index + 1):
            cur = cur.next
        return cur

    def isNotPositionIndex(self, index: int) -> bool:
        return not self.isPositionIndex(index)

    def isPositionIndex(self, index: int) -> bool:
        return index >= 0 and index < self.size

File: dsa-learning/dsa-java-code/src/main/java/com/dsa/leetcode/labuladong2/list/P707_DesignLinkedList.java
package com.dsa.leetcode.labuladong2.list;

public class P707_DesignLinkedList {

    class MyLinkedList {
        Node head;
        Node tail;
        int size;

        public MyLinkedList() {
            head = new Node();
            tail = new Node();
            head.next = tail;
            tail.prev = head;
            size = 0;
        }

        public int get(int index) {
            if (isNotPositionIndex(index)) {
                return -1;
            }
            Node node = getNode(index);
            return node.val;
        }

        public void addAtHead(int val) {
            addAtIndex(0, val);
        }

        public void addAtTail(int val) {
            addAtIndex(size, val);
        }

        public void addAtIndex(int index, int val) {
            if (isPositionIndex(index)) {
                Node cur = getNode(index);
                Node prev = cur.prev;
                Node newNode = new Node(val);
                prev.next = newNode;
                cur.prev = newNode;
                newNode.prev = prev;
                newNode.next = cur;
                size++;
            }
            if (index == size) {
                Node newNode = new Node(val);
                Node prev = tail.prev;
                prev.next = newNode;
                tail.prev = newNode;
                newNode.prev = prev;
                newNode.next = tail;
                size++;
            }
        }

        public void deleteAtIndex(int index) {
            if (isPositionIndex(index)) {
                Node cur = getNode(index);
                Node prev = cur.prev;
                Node next = cur.next;
                prev.next = next;
                next.prev = prev;
                size--;
            }
        }

        public Node getNode(int index) {
            Node cur = head;
            for (int i = 0; i <= index; i++) {
                cur = cur.next;
            }
            return cur;
        }

        public void checkElementIndex(int index) {
            if (isNotPositionIndex(index)) {
                throw new IndexOutOfBoundsException();
            }
        }

        public boolean isPositionIndex(int index) {
            return index >= 0 && index < size;
        }

        public boolean isNotPositionIndex(int index) {
            return !isPositionIndex(index);
        }

        static class Node {

            int val;
            Node prev;
            Node next;

            public Node() {

            }

            public Node(int val) {
                this.val = val;
            }

            public Node(int val, Node prev, Node next) {
                this.val = val;
                this.prev = prev;
                this.next = next;
            }
        }
    }

}
